shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/antlr4/imports/oracle/DMLStatement.g4

```diff
@@ -121,7 +121,223 @@ unionClause
     ;
 
 queryBlock
-    : SELECT duplicateSpecification? projections fromClause? whereClause? groupByClause? havingClause?
+    : withClause? SELECT duplicateSpecification? projections fromClause? whereClause? groupByClause? havingClause?
+    ;
+
+withClause
+    : WITH plsqlDeclarations? ((subqueryFactoringClause | subavFactoringClause) (COMMA_ (subqueryFactoringClause | subavFactoringClause))*)?
+    ;
+
+plsqlDeclarations
+    : (functionDeclaration | procedureDeclaration)+
+    ;
+
+functionDeclaration
+    : functionHeading ((DETERMINISTIC | PIPELINED | PARALLEL_ENABLE | RESULT_CACHE)+)?
+    ;
+
+functionHeading
+    : FUNCTION functionName (LP_ parameterDeclaration (SQ_ parameterDeclaration)* RP_)? RETURN dataType
+    ;
+
+parameterDeclaration
+    : parameterName ((IN? dataType ((COLON_ EQ_ | DEFAULT) expr)?) | (IN? OUT NOCOPY? dataType))?
+    ;
+
+procedureDeclaration
+    : procedureHeading procedureProperties
+    ;
+
+procedureHeading
+    : PROCEDURE procedureName (LP_ parameterDeclaration (SQ_ parameterDeclaration)* RP_)?
+    ;
+
+procedureProperties
+    : (accessibleByClause | defaultCollationClause | invokerRightsClause)*
+    ;
+
+accessibleByClause
+    : ACCESSIBLE BY LP_ accessor (COMMA_ accessor)* RP_
+    ;
+
+accessor
+    : unitKind? unitName
+    ;
+
+unitKind
+    : FUNCTION | PROCEDURE | PACKAGE | TRIGGER | TYPE
+    ;
+
+defaultCollationClause
+    : DEFAULT COLLATION collationOption
+    ;
+
+collationOption
+    : USING_NLS_COMP
+    ;
+
+invokerRightsClause
+    : AUTHID (CURRENT_USER | DEFINER)
+    ;
+
+subqueryFactoringClause
+    : queryName (LP_ alias (COMMA_ alias)* RP_)? AS LP_ selectSubquery RP_ searchClause? cycleClause? 
+    ;
+
+searchClause
+    : SEARCH (DEPTH | BREADTH) FIRST BY (alias (ASC | DESC)? (NULLS FIRST | NULLS LAST)?) (COMMA_ (alias (ASC | DESC)? (NULLS FIRST | NULLS LAST)?))* SET orderingColumn
+    ;
+
+cycleClause
+    : CYCLE alias (COMMA_ alias)* SET alias TO cycleValue DEFAULT noCycleValue
+    ;
+
+subavFactoringClause
+    : subavName ANALYTIC VIEW AS LP_ subavClause RP_
+    ;
+
+subavClause
+    : USING baseAvName hierarchiesClause? filterClauses? addCalcsClause?
+    ;
+
+hierarchiesClause
+    : HIERARCHIES LP_ ((alias DOT_)? alias (COMMA_ (alias DOT_)? alias)*)? RP_
+    ;
+
+filterClauses
+    : FILTER FACT LP_ filterClause (COMMA_ filterClause)* RP_
+    ;
+
+filterClause
```

Comment:
![image](https://user-images.githubusercontent.com/22066046/120728991-a72fac00-c510-11eb-9b2b-b555d908fa17.png)

Reply:
![image](https://user-images.githubusercontent.com/22066046/120872435-58961680-c5d1-11eb-9c09-4c09052b054b.png)

@ThanoshanMV Is it suitable that 'predicate' is written as an 'expr'?

Reply:
I earlier put `predicate` for `filterClause`, it parsed them correctly for two examples in the above image but there's another example SQL statement that defines a transitory analytic view using a filter clause:

![FilterFactClause](https://user-images.githubusercontent.com/48581379/120881821-796e6400-c5f1-11eb-9ea1-0fd59764edd6.png)

The above statement is failed to parse correctly when I specified `predicate`:

![PredicateRuleFailure](https://user-images.githubusercontent.com/48581379/120881877-dcf89180-c5f1-11eb-9aa3-3b4bd6899f24.png)

Then I changed `predicate` to `expr`:

![ExprRuleSuccess](https://user-images.githubusercontent.com/48581379/120881922-247f1d80-c5f2-11eb-966f-9a5dae1dfb99.png)

Reply:
Because the definition of the predicate is different from the definition of the expr.
```
predicate
    : bitExpr NOT? IN subquery
    | bitExpr NOT? IN LP_ expr (COMMA_ expr)* RP_
    | bitExpr NOT? BETWEEN bitExpr AND predicate
    | bitExpr NOT? LIKE simpleExpr (ESCAPE simpleExpr)?
    | bitExpr
    ;
``

Reply:
Maybe you should add some rules for the predicate.

Reply:
I've added a new rule in `predicate` as `bitExpr NOT? IN LP_ expr (COMMA_ expr)* RP_ AND predicate` that matches with the above example:

![PredicateRuleInFilterClause](https://user-images.githubusercontent.com/48581379/121223122-d5144800-c8a4-11eb-879e-a3cbf32b1325.png)